### ISSUE TITLE
Binding OfflineAudioContext

### ISSUE THREAD

**Issue by bthj** [ircam-ismm/node-web-audio-api#4]
Hi - super interesting project you have here!

It seems that `OfflineAudioContext` hasn't yet been mapped ([web-audio-api-rs](https://github.com/orottier/web-audio-api-rs) seems to implement it).  If I'd try to add that binding, should it be something similar to what has already been done for `AudioContext`?
https://github.com/ircam-ismm/node-web-audio-api/blob/main/index.js#L91

**Comment by b-ma** [comment]
Hi, thanks!

No indeed, still have some work to reach the `web-audio-api-rs` current state. Actually, implementing the bindings for the offline context should not be much complicated I think. Starting from https://github.com/ircam-ismm/node-web-audio-api/blob/main/generator/templates/audio_context.tmpl.rs it should be mostly duplicating the file, modifying the constructor to handle the constructor signature and removing the extra online context methods (resume, suspend, etc.)

The most complicated thing to grasp I guess is that all the rust source files for the bindings located in `src` are dynamically generated from the files and scripts located in the `generator` directory, so the general workflow is:
- create / modify the files in `generator`
- generate the rust source files from the templates with `npm run generate`
- build the native node module from the generated rust sources with`npm run build`

Let me know if you want to have a try and/or need some help, I can also have a shot on this if you don't feel confortable with the source code.

**Comment by bthj** [comment]
Thanks for the insights.  Maybe I should have a try and see how far I get.  Will post back :)

**Comment by b-ma** [comment]
ok cool, let me know!

**Comment by bthj** [comment]
I've taken a stab at adding this but it's not quite building yet.  Here's what I've done so far:

https://github.com/ircam-ismm/node-web-audio-api/compare/main...bthj:node-web-audio-api:feature/offline_audio_context

The constructor signature is different (offline with 3 arguments while online with one options arg) and how that's handled doesn't seem to produce errors.  But an attempt at implementing a binding for the `start_rendering` results in errors.  There I started by copying the implementation for `decode_audio_data` as it's also handling an audio buffer, but the signature is a bit different, where `decode_audio_data` returns a `Result<Audiobuffer ... >` but `start_rendering` returns `AudioBuffer` - and that's causing a build error I currently don't know how to handle due to unfamiliarity with napi-rs.  It says 
`cannot use the '?' operator in a function that returns 'AudioBuffer'`
at 
`let napi_obj = ctx.env.unwrap::<NapiOfflineAudioContext>(&js_this)?;`

Hopefully there are just minor changes required to have this building.  I'll attach the current build error log, in case it helps pushing this in the right direction:
[build-error_2022-10-30.txt](https://github.com/ircam-ismm/node-web-audio-api/files/9897010/build-error_2022-10-30.txt)

**Comment by b-ma** [comment]
Hey, thanks I managed to make this work but there was actually other issues hiding after that point...

According to your problem, I ended up with something like that (not being very good with Rust, and considering `napi-rs` mostly as black magic, I hope I won't tell too much idiocies):

```rs
#[js_function]
// the function here should return a Result<JsObject>, not a raw rust AudioBuffer
fn start_rendering(ctx: CallContext) -> Result<JsObject> {
    let js_this = ctx.this_unchecked::<JsObject>();
    let napi_obj = ctx.env.unwrap::<NapiOfflineAudioContext>(&js_this)?;
    // really not sure this is the best way, but we need ownership over the rust `audio_context`
    // here because of the `start_rendering` signature and this the only way I found to retrieve 
    // it from the raw pointer napi-rs is giving us
    let owned = unsafe { Box::from_raw(napi_obj) };
    let context = owned.consume();

    // as start_rendering_sync returns an `AudioBuffer` and not a `Result<AudioBuffer>` 
    // we don't need the match branches
    let audio_buffer = context.start_rendering_sync();

    // I think there was no changes after that point
    let store_ref: &mut napi::Ref<()> = ctx.env.get_instance_data()?.unwrap();
    let store: JsObject = ctx.env.get_reference_value(store_ref)?;
    let ctor: JsFunction = store.get_named_property("AudioBuffer")?;
    let mut options = ctx.env.create_object()?;
    options.set("__internal_caller__", ctx.env.get_null())?;

    let js_audio_buffer = ctor.new_instance(&[options])?;
    let napi_audio_buffer = ctx.env.unwrap::<NapiAudioBuffer>(&js_audio_buffer)?;
    napi_audio_buffer.populate(audio_buffer);

    Ok(js_audio_buffer)
}
```

But then there was a big problem with **all the nodes** that were waiting for an AudioContext and not an OfflineAudioContext... I managed to make it work, maybe not with the most elegant solution but that works.

Maybe, you could open a PR so that I can try to push my changes on your branch and continue iterating?

**Comment by bthj** [comment]
Thanks 🙏  PR has been opened :)

**Comment by b-ma** [comment]
done in #5

**Comment by b-ma** [comment]
landed in version 0.4.0
